### ISSUE TITLE
Can't manage to have Container with `withProps: true` working since #346

### ISSUE THREAD

**Issue by bcharbonnier** [facebook/flux#382]
Hi everyone!

I noticed that since the introduction of #346 Containers created with `withProps: true` are not working anymore, or at least I did not managed to get them working the way I used to.

**Disclaimer**: it's a long read, but I try to explain it as clear as I can. For the one in a hurry, I try to give an explanation at the end of the description:

Let me try to reproduce it with a minimal setup for create-react-app (I don't know how to Flux in Plunker, or Codepen, or ...)
My use case is quite simple (or maybe I am doing something totally wrong, then please tell me):
I have 2 components with a Parent-Child relationship, both are Containers, the child is a Container that get a prop injected to calculate her state.
My two components are listening to 2 different stores. I then update these stores one after the other with proper delays of 2500ms and 5000ms.

I have setup a default value for my injected prop, so you could see what it does without or with `withProps: true` on `NameContainer` that is causing the Invariant Violation

So here is my scenario:

```bash
create-react-app flux-with-props
cd flux-with-props
npm install flux immutable
```
Replace the content of App.js

```jsx
import React, { Component } from 'react';
import { Map } from "immutable";

import { Dispatcher as FluxDispatcher } from "flux";
import { ReduceStore, Container } from "flux/utils";

const Dispatcher = new FluxDispatcher();

const NameStore = new (class extends ReduceStore {
  getInitialState() {
    return new Map({ name: "Julian" });
  }

  reduce(state, payload) {
    switch(payload.type) {
      case "CHANGE_NAME":
        return state.set("name", payload.name);
      default:
        return state;
    }
  }

  getName(mood) {
    return `${this.getState().get("name")} is ${mood}`;
  }
})(Dispatcher);

const InfoStore = new (class extends ReduceStore {
  getInitialState() {
    return new Map();
  }

  reduce(state, payload) {
    switch(payload.type) {
      case "UPDATE_MOOD":
        return state.set("info", {mood: payload.mood});
      default:
        return state;
    }
  }

  getInfo(mood = "sad") {
    return this.getState().get("info", {mood});
  }
})(Dispatcher);

class Name extends Component {
  static getStores() {
    return [NameStore];
  }

  static calculateState(prevState, props) {
    const mood = (props && props.info && props.info.mood);
    return {
      name: NameStore.getName(mood || "neither sad nor happy")
    }
  }

  render() {
    return (
      <div>{this.state.name}</div>
    )
  }
}

const NameContainer = Container.create(Name, { withProps: false });

class Wrapper extends Component {
  static getStores() {
    return [InfoStore];
  }

  static calculateState() {
    return {
      info: InfoStore.getInfo()
    }
  }

  render() {
    return (
      <div>
        <h3>Information</h3>
        <NameContainer info={this.state.info} />
      </div>
    );
  }
}

const WrapperContainer = Container.create(Wrapper);

setTimeout(() => {
  Dispatcher.dispatch({
    type: "UPDATE_MOOD",
    mood: "happy"
  });
}, 5000);

setTimeout(() => {
  Dispatcher.dispatch({
    type: "CHANGE_NAME",
    name: "Patrick"
  });
}, 2500);


class App extends Component {
  render() {
    return (
      <div className="App">
        <WrapperContainer />
      </div>
    );
  }
}

export default App;
```

```bash
npm run start
```

You should see

_Julian is neither sad nor happy_

_Patrick is neither sad nor happy_

Now activate `withProps: true` on NameContainer (line 66).
We will then inject a prop on it that default value is still _neither sad nor happy_ but we also dispatch a change to _happy_ after a delay.

We then should see

_Julian is neither sad nor happy_

_Patrick is neither sad nor happy_

_Patrick is happy_

But the last state never occurs, due to the invariant check.

---

**My debugging session lead me here:**
At one point the `FluxContainerSubscriptions` will trigger a reset on the underlying `FluxStoreGroup` that will trigger a `Dispatcher.unregister/register`. But if we are here, it's because we are reacting to a prop change that might be triggered by a dispatch !

There it is... a dispatch that trigger an unregister/register. Boom. Invariant :(

Am I really doing something wrong here ? 2 components encapsulated, 2 stores, updates... Sounds pretty normal to me...

Thanks for the help, and really sorry for the long reading !!

**Comment by kyldvs** [comment]
Thanks for the report, will investigate shortly. I think we may need to remove the invariant or add in a workaround, whichever seems cleanest.

**Comment by alexeyraspopov** [comment]
@kyldvs, maybe we can get rid of the logic which resets listeners each time before calling `getStores()`? I never saw a case where I'd put some logic in my `getStores()` methods. It's always just a list of all stores that are used by any condition in `calculateState()`.

**Comment by kyldvs** [comment]
That was a particular feature that was requested in order to support passing stores down via props.

I'm just going to remove these invariants, I don't think they catch a common issue or something that is necessarily an anti-pattern.

**Comment by kyldvs** [comment]
Published 3.1.0 which should fix the issue. Let me know if it doesn't!

**Comment by bcharbonnier** [comment]
Thanks @kyldvs ! Works like a charm 👍
